src/contour/Audio.cpp

```diff
@@ -0,0 +1,115 @@
+    return std::isgreater(x, 1) ? -1 : 1;
+}
+
+auto createMusicalNote(double volume, double duration, double frequency) noexcept
+{
+    duration = duration / 32.0;
+    volume /= 7;
+    std::vector<std::int16_t> buffer;
+    buffer.reserve(static_cast<size_t>(std::ceil(duration * SAMPLE_RATE)));
+    for (double i = 0; i < duration * SAMPLE_RATE; ++i)
+    {
+        auto sample = static_cast<int16_t>(0x7fff * volume * square_wave(frequency / SAMPLE_RATE * i * 2));
+        buffer.push_back(sample);
+    }
+    return buffer;
+}
+
+} // namespace
+
+Audio::Audio()
+{
+    QAudioFormat f;
+    f.setSampleRate(44100);
+    f.setChannelCount(1);
+    f.setSampleSize(16);
+    f.setCodec("audio/pcm");
+    f.setByteOrder(QAudioFormat::LittleEndian);
+    f.setSampleType(QAudioFormat::SignedInt);
+    QAudioDeviceInfo info(QAudioDeviceInfo::defaultOutputDevice());
+    if (!info.isFormatSupported(f))
+    {
+        errorlog()("Default output device doesn't support 16 Bit signed integer PCM");
+        return;
+    }
+    audio = std::make_unique<QAudioOutput>(f);
+    audio->moveToThread(&soundThread_);
+    connect(audio.get(), &QAudioOutput::stateChanged, this, &Audio::handleStateChanged);
+    qRegisterMetaType<std::vector<int>>();
+    connect(this, &Audio::play, this, &Audio::handlePlayback, Qt::ConnectionType::QueuedConnection);
+    soundThread_.start();
+}
+
+Audio::~Audio()
+{
+    soundThread_.quit();
+    soundThread_.wait();
+}
+
+void Audio::fillBuffer(int volume, int duration, crispy::span<const int> notes)
+{
+    for (auto i: notes)
+    {
+        auto b = createMusicalNote(volume, duration, i);
+        byteArray_.append(reinterpret_cast<const char*>(b.data()), static_cast<int>(2 * b.size()));
+    }
+}
+
+void Audio::handlePlayback(int volume, int duration, std::vector<int> const& notes)
+{
+    if (audio->state() == QAudio::State::ActiveState)
+    {
+        fillBuffer(volume, duration, crispy::span(notes.data(), notes.size()));
+        return;
+    }
+    fillBuffer(volume, duration, crispy::span(notes.data(), notes.size()));
+    audioBuffer_.setBuffer(&byteArray_);
+    audioBuffer_.open(QIODevice::ReadWrite);
+    if (audio)
+        audio->start(&audioBuffer_);
+}
+
+void Audio::handleStateChanged(QAudio::State state)
+{
+    switch (state)
+    {
+        case QAudio::IdleState:
+            audio->stop();
+            audioBuffer_.close();
+            byteArray_.clear();
+            break;
+
+        case QAudio::StoppedState:
+            if (audio->error() != QAudio::NoError)
+            {
+                errorlog()("Audio playback stopped: {}", audio->error());
+            }
+            break;
+        default: break;
+    }
+}
+
+std::vector<std::int16_t> Audio::createMusicalNote(double volume, int duration, int note_) noexcept
+{
+    Require(static_cast<int>(note_) >= 0 && static_cast<int>(note_) < 26);
+    double frequency = (note_ != 0) * 440.0 * std::pow(2, (note_ + 2) / 12.0);
```

Comment:
Very nice. Looks much much shorter than the initial version.

EDIT: deleted all my text. 

I did a little test in [compiler explorer](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:1,endLineNumber:4,positionColumn:1,positionLineNumber:4,selectionStartColumn:1,selectionStartLineNumber:4,startColumn:1,startLineNumber:4),source:'%23include+%3Ccmath%3E%0A%23include+%3Cclimits%3E%0A%0Aint+branching_cpp(double+note)+%7B%0A++++if+(note+%3C+std::numeric_limits%3Cdouble%3E::epsilon())%0A++++++++return+0%3B%0A++++else%0A++++++++return+440.0+*+std::pow(2,+(note+%2B+2)+/+12.0)%3B%0A%7D%0A%0Aint+branchless_cpp(double+note)+%7B%0A++++return+(note+!!%3D+0)+*+440.0+*+std::pow(2,+(note+%2B+2)+/+12.0)%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:g121,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-O3+-msse2',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1,tree:'1'),l:'5',n:'0',o:'x86-64+gcc+12.1+(C%2B%2B,+Editor+%231,+Compiler+%231)',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4).

Turns out that using an explicit branch is (from simply looking at it) shorter and uses only 2 basic blocks whereas the branchless version uses three. Not sure what's faster. I'm fine with both versions. :-)